### ISSUE TITLE
SLPVectorizer crashes on valid code at -Os and above on x86_64-linux-gnu: Assertion `VecTy == FinalVecTy && "No reused scalars expected for broadcast."' failed

### ISSUE THREAD

**Issue by zhendongsu** [llvm/llvm-project#52756]
```
[508] % clangtk -v
clang version 14.0.0 (https://github.com/llvm/llvm-project.git cce93b339724a34928f0630fa8ed6b09b4ba753b)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /local/suz-local/opfuzz/bin
Found candidate GCC installation: /usr/lib/gcc/i686-linux-gnu/8
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Candidate multilib: x32;@mx32
Selected multilib: .;@m64
[509] % 
[509] % clangtk -O1 small.c; ./a.out
[510] % 
[510] % clangtk -Os small.c
clang-14: /local/suz-local/software/clangbuild/llvm-project/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:4740: llvm::InstructionCost llvm::slpvectorizer::BoUpSLP::getEntryCost(const llvm::slpvectorizer::BoUpSLP::TreeEntry*, llvm::ArrayRef<llvm::Value*>): Assertion `VecTy == FinalVecTy && "No reused scalars expected for broadcast."' failed.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /local/suz-local/software/local/clang-trunk/bin/clang-14 -cc1 -triple x86_64-unknown-linux-gnu -emit-obj --mrelax-relocations -disable-free -clear-ast-before-backend -main-file-name small.c -mrelocation-model static -mframe-pointer=none -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -funwind-tables=2 -target-cpu x86-64 -tune-cpu generic -debugger-tuning=gdb -fcoverage-compilation-dir=/local/suz-local/software/emitesting/bugs/20211216-clangtk-m64-O3-Weverything-pipe-fPIC-build-143706/delta -resource-dir /local/suz-local/software/local/clang-trunk/lib/clang/14.0.0 -I /usr/local/include/csmith -I /local/suz-local/software/local/include -internal-isystem /local/suz-local/software/local/clang-trunk/lib/clang/14.0.0/include -internal-isystem /usr/local/include -internal-isystem /usr/lib/gcc/x86_64-linux-gnu/7.5.0/../../../../x86_64-linux-gnu/include -internal-externc-isystem /usr/include/x86_64-linux-gnu -internal-externc-isystem /include -internal-externc-isystem /usr/include -Os -fdebug-compilation-dir=/local/suz-local/software/emitesting/bugs/20211216-clangtk-m64-O3-Weverything-pipe-fPIC-build-143706/delta -ferror-limit 19 -fgnuc-version=4.2.1 -fcolor-diagnostics -vectorize-loops -vectorize-slp -faddrsig -D__GCC_HAVE_DWARF2_CFI_ASM=1 -o /tmp/small-09b7fb.o -x c small.c
1.	<eof> parser at end of file
2.	Optimizer
 #0 0x000055865e4c287f PrintStackTraceSignalHandler(void*) Signals.cpp:0:0
 #1 0x000055865e4c005d SignalHandler(int) Signals.cpp:0:0
 #2 0x00007f6e9ee86980 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12980)
 #3 0x00007f6e9db37fb7 raise /build/glibc-S9d2JN/glibc-2.27/signal/../sysdeps/unix/sysv/linux/raise.c:51:0
 #4 0x00007f6e9db39921 abort /build/glibc-S9d2JN/glibc-2.27/stdlib/abort.c:81:0
 #5 0x00007f6e9db2948a __assert_fail_base /build/glibc-S9d2JN/glibc-2.27/assert/assert.c:89:0
 #6 0x00007f6e9db29502 (/lib/x86_64-linux-gnu/libc.so.6+0x30502)
 #7 0x000055865e714bac llvm::slpvectorizer::BoUpSLP::getEntryCost(llvm::slpvectorizer::BoUpSLP::TreeEntry const*, llvm::ArrayRef<llvm::Value*>) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x3c63bac)
 #8 0x000055865e735762 llvm::slpvectorizer::BoUpSLP::getTreeCost(llvm::ArrayRef<llvm::Value*>) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x3c84762)
 #9 0x000055865e737d99 llvm::SLPVectorizerPass::tryToVectorizeList(llvm::ArrayRef<llvm::Value*>, llvm::slpvectorizer::BoUpSLP&, bool) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x3c86d99)
#10 0x000055865e73e19a llvm::SLPVectorizerPass::vectorizeChainsInBlock(llvm::BasicBlock*, llvm::slpvectorizer::BoUpSLP&) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x3c8d19a)
#11 0x000055865e74215c llvm::SLPVectorizerPass::runImpl(llvm::Function&, llvm::ScalarEvolution*, llvm::TargetTransformInfo*, llvm::TargetLibraryInfo*, llvm::AAResults*, llvm::LoopInfo*, llvm::DominatorTree*, llvm::AssumptionCache*, llvm::DemandedBits*, llvm::OptimizationRemarkEmitter*) (.part.2338) SLPVectorizer.cpp:0:0
#12 0x000055865e74316b llvm::SLPVectorizerPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x3c9216b)
#13 0x000055865f729351 llvm::detail::PassModel<llvm::Function, llvm::SLPVectorizerPass, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Function> >::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x4c78351)
#14 0x000055865dc42c31 llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function> >::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x3191c31)
#15 0x000055865c127061 llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function> >, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Function> >::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x1676061)
#16 0x000055865dc4263a llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x319163a)
#17 0x000055865c1279c1 llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::PreservedAnalyses, llvm::AnalysisManager<llvm::Module> >::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x16769c1)
#18 0x000055865dc40d7d llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module> >::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x318fd7d)
#19 0x000055865e7ef6a0 (anonymous namespace)::EmitAssemblyHelper::RunOptimizationPipeline(clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >&, std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile> >&) (.constprop.827) BackendUtil.cpp:0:0
#20 0x000055865e7f2632 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream> >) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x3d41632)
#21 0x000055865f54e90b clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x4a9d90b)
#22 0x000055866042e2f9 clang::ParseAST(clang::Sema&, bool, bool) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x597d2f9)
#23 0x000055865f54d418 clang::CodeGenAction::ExecuteAction() (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x4a9c418)
#24 0x000055865ee65a71 clang::FrontendAction::Execute() (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x43b4a71)
#25 0x000055865ee01d1a clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x4350d1a)
#26 0x000055865ef34d23 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x4483d23)
#27 0x000055865be77d4c cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x13c6d4c)
#28 0x000055865be73119 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#29 0x000055865bd9c001 main (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x12eb001)
#30 0x00007f6e9db1abf7 __libc_start_main /build/glibc-S9d2JN/glibc-2.27/csu/../csu/libc-start.c:344:0
#31 0x000055865be72c8a _start (/local/suz-local/software/local/clang-trunk/bin/clang-14+0x13c1c8a)
clang-14: error: unable to execute command: Aborted
clang-14: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 14.0.0 (https://github.com/llvm/llvm-project.git cce93b339724a34928f0630fa8ed6b09b4ba753b)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /local/suz-local/opfuzz/bin
clang-14: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang-14: note: diagnostic msg: /tmp/small-ee660f.c
clang-14: note: diagnostic msg: /tmp/small-ee660f.sh
clang-14: note: diagnostic msg: 

********************
[511] % 
[511] % cat small.c
int printf(const char *, ...);
int a, *b, *c, d, e, f;
int main() {
  int g = 0, h, i = 0, j, k, o;
  if (a) {
    for (; *b; (*b)++)
      ;
    h = e / g;
    i = printf("%d", 1);
    int l = g;
    if (d)
      printf("%d", h);
    j = g;
    if (g)
      h = l >> g;
    if (!h) {
      printf("%d", g);
      goto L1;
    }
    while (j)
      c = &g;
  }
  int n = g;
L2:
  if (i) {
    k = j;
    o = f;
  L1:
    if (g && i)
      goto L2;
    if (k) {
      printf("%d", g);
      goto L1;
    }
    while (g)
      printf("0");
    goto L2;
  }
  if (n)
    while (o)
      ;
}
```

From Compiler Explorer: https://godbolt.org/z/v19nf97ez

**Comment by fhahn** [comment]
cc @alexey-bataev

**Comment by alexey-bataev** [comment]
https://reviews.llvm.org/D115939

**Comment by alexey-bataev** [comment]
Fixed in fcaf290d0278bb83387e1a1d972c55e08b8c40e3
